src/dot.ts

```diff
@@ -5,30 +5,44 @@
 
 import { Vex } from './vex';
 import { Modifier } from './modifier';
+import { Note } from './note';
+import { StaveNote } from './stavenote';
 
 export class Dot extends Modifier {
-  static get CATEGORY() {
+  protected note?: Note;
+  protected radius: number;
+  protected dot_shiftY: number;
+
+  static get CATEGORY(): string {
     return 'dots';
   }
 
   // Arrange dots inside a ModifierContext.
-  static format(dots, state) {
+  static format(
+    dots: Dot[],
+    state: {
+      right_shift: number;
+      left_shift: number;
+      text_line: number;
+      top_text_line: number;
+    }
+  ): boolean {
     const right_shift = state.right_shift;
     const dot_spacing = 1;
 
     if (!dots || dots.length === 0) return false;
 
     const dot_list = [];
-    const max_shift_map = {};
+    const max_shift_map: Record<string, number> = {};
     for (let i = 0; i < dots.length; ++i) {
       const dot = dots[i];
       const note = dot.getNote();
 
       let props;
       let shift;
 
-      // Only StaveNote has .getKeyProps()
-      if (typeof note.getKeyProps === 'function') {
+      // If it's a StaveNote
+      if (note instanceof StaveNote) {
         props = note.getKeyProps()[dot.getIndex()];
         shift = note.getRightDisplacedHeadPx();
       } else {
```

Comment:
Good place to explicitly check for `TabNote` and throw an exception in a default `else`.

Reply:
done